build.rs

```diff
@@ -134,52 +215,148 @@ fn fix_config_map(mut config_map: HashMap<String, String>) -> HashMap<String, St
     config_map
 }
 
-fn load_cross_compile_info() -> Result<(InterpreterConfig, HashMap<String, String>)> {
-    let python_include_dir = env::var("PYO3_CROSS_INCLUDE_DIR")?;
-    let python_include_dir = Path::new(&python_include_dir);
+fn parse_script_output(output: &str) -> HashMap<String, String> {
+    output
+        .lines()
+        .filter_map(|line| {
+            let mut i = line.splitn(2, ' ');
+            Some((i.next()?.into(), i.next()?.into()))
+        })
+        .collect()
+}
 
-    let patchlevel_defines = parse_header_defines(python_include_dir.join("patchlevel.h"))?;
+/// Parse sysconfigdata file
+///
+/// The sysconfigdata is basically a dictionary, and since we can't really use this library to read
+/// it for us, egg and chicken type thing, we parse it with some regex. Here there are two regex's.
+/// The first one is to match an entry in the dictionary (`entry_re`). Then when the entry is on
+/// multiple lines we use the second regex to capture the additional string entry and add it to the
+/// previously captured key. We detect if this is a multi line entry by checking if the last capture
+/// group contains either `,` or `}`.
+fn parse_sysconfigdata(config_path: impl AsRef<Path>) -> Result<HashMap<String, String>> {
+    let mut script = fs::read_to_string(config_path)?;
+    script += r#"
+print("version_major", build_time_vars["VERSION"][0])  # 3
+print("version_minor", build_time_vars["VERSION"][2])  # E.g., 8
+if "WITH_THREAD" in build_time_vars:
+    print("WITH_THREAD", build_time_vars["WITH_THREAD"])
+if "Py_TRACE_REFS" in build_time_vars:
+    print("Py_TRACE_REFS", build_time_vars["Py_TRACE_REFS"])
+if "COUNT_ALLOCS" in build_time_vars:
+    print("COUNT_ALLOCS", build_time_vars["COUNT_ALLOCS"])
+if "Py_REF_DEBUG" in build_time_vars:
+    print("Py_REF_DEBUG", build_time_vars["Py_REF_DEBUG"])
+print("Py_DEBUG", build_time_vars["Py_DEBUG"])
+print("Py_ENABLE_SHARED", build_time_vars["Py_ENABLE_SHARED"])
+print("LDVERSION", build_time_vars["LDVERSION"])
+print("SIZEOF_VOID_P", build_time_vars["SIZEOF_VOID_P"])
+"#;
+    let output = run_python_script(&find_interpreter()?, &script)?;
+
+    Ok(parse_script_output(&output))
+}
+
+fn starts_with(entry: &DirEntry, pat: &str) -> bool {
+    let name = entry.file_name();
+    name.to_string_lossy().starts_with(pat)
+}
+fn ends_with(entry: &DirEntry, pat: &str) -> bool {
+    let name = entry.file_name();
+    name.to_string_lossy().ends_with(pat)
+}
 
-    let major = match patchlevel_defines
-        .get("PY_MAJOR_VERSION")
-        .map(|major| major.parse::<u8>())
-    {
-        Some(Ok(major)) => major,
-        Some(Err(e)) => bail!("Failed to parse PY_MAJOR_VERSION: {}", e),
-        None => bail!("PY_MAJOR_VERSION undefined"),
+fn find_sysconfigdata(path: impl AsRef<Path>, cross: &CrossCompileConfig) -> Option<PathBuf> {
+    for f in fs::read_dir(path).expect("Path does not exist") {
```

Comment:
This might need adjusting just very slightly so that it works from just the "lib" dir on unix platforms? See https://github.com/PyO3/pyo3/issues/1110#issuecomment-675179064

Reply:
Yeah so from what I understand is the fact that if we link to a python installation we get a structure like so

```
/usr/lib/libpython3.Y.so
/usr/lib/python3.Y/_sysconfigdata*.py
```

The issue now is matching against the right python version, so for example if python 3.6, 3.7 and 3.8 are installed on the system, this will simply match the first one found . We might need a new environment variable just to define the version of python if the first one found is not the wanted version.

Reply:
As it seems like this script will now already work correctly if pointed at `/usr/lib/python3.Y`, maybe we can just document that instead of adding a new variable?

Along that line: to avoid suprises I suggest if multiple python versions are matched, pyo3 should bail and ask for a more specific `PYO3_CROSS_LIB_DIR` to be set. (This seems better to me than silently picking the first one.)

Reply:
The problem would arise when needing to link to `libpython` as it would not always be in the lib dir. For example, the raspberry pi seems to have it within the `python3.Y` directory, how ever on Manjaro, only the static library is within that directory and the shared library is top level under `/usr/lib`. So if the user needs to link to `libpython.so` such as when building for android or making an executable, we wouldn't be able to find it.

What I'm thinking of doing is keeping the recursive search, adding the search for the `python3.Y` directory and instead of returning immediately when a match is found, keeping a record of the locations found. Like that we can bail with a useful error message, or we get all the necessary paths. I would still add an optional environment variable to set the version when needed. Since if I have a system with the following file placements, I will not be the only one.

```
/usr/lib/libpython3.7.so
/usr/lib/python3.7/_sysconfigdata*.py
/usr/lib/libpython3.8.so
/usr/lib/python3.8/_sysconfigdata*.py
/usr/lib/python3.8/config-3.8-x86_64-linux-gnu/libpython3.8.a
```

Reply:
see commit 7beb272 for the application of this, I noticed that linking still works even if the `libpython` is deep in the directory. Such as when the DSO is only at `/usr/lib/python3.8/config-3.8-x86_64-linux-gnu/libpython3.8.so` and you set `PYO3_CROSS_LIB_DIR=/usr/lib`